numpyro/infer/hmc_gibbs.py

```diff
@@ -149,3 +154,188 @@ def potential_fn(z_gibbs, z_hmc):
+
+        def potential_fn(z_discrete):
+            model_kwargs["_gibbs_sites"] = z_discrete
+            return potential_energy(_wrap_model(model), model_args, model_kwargs, z_hmc, enum=enum)
+
+        # get support_sizes of gibbs_sites
+        support_sizes_flat, _ = ravel_pytree({k: support_sizes[k] for k in gibbs_sites})
+        num_discretes = support_sizes_flat.shape[0]
+
+        rng_key, rng_permute = random.split(rng_key)
+        idxs = random.permutation(rng_key, jnp.arange(num_discretes))
+
+        def body_fn(i, val):
+            idx, support_size = idxs[i], support_sizes_flat[i]
+            return _discrete_step(*val, potential_fn=potential_fn, idx=idx, support_size=support_size)
+
+        init_val = (rng_key, gibbs_sites, potential_fn(gibbs_sites))
+        _, gibbs_sites, _ = fori_loop(0, num_discretes, body_fn, init_val)
+        return gibbs_sites
+
+    return gibbs_fn
+
+
+def subsample_gibbs_fn(model, *model_args, **model_kwargs):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    Returns a gibbs_fn to be used in :class:`HMCGibbs`, which works for subsampling
+    statements using :class:`~numpyro.plate` primitive. This implements the Algorithm 1
+    of reference [1] but uses a naive estimation (without control variates) of log likelihood,
+    hence might incur a high variance.
+
+    **References:**
+
+    1. *Hamiltonian Monte Carlo with energy conserving subsampling*,
+       Dang, K. D., Quiroz, M., Kohn, R., Minh-Ngoc, T., & Villani, M. (2019)
+    2. *Speeding Up MCMC by Efficient Data Subsampling*,
+       Quiroz, M., Kohn, R., Villani, M., & Tran, M. N. (2018)
+
+    **Example**
+
+    .. doctest::
+
+        >>> from jax import random
+        >>> import jax.numpy as jnp
+        >>> import numpyro
+        >>> import numpyro.distributions as dist
+        >>> from numpyro.infer import MCMC, NUTS, HMCGibbs, subsample_gibbs_fn
+        ...
+        >>> def model(data):
+        ...     x = numpyro.sample("x", dist.Normal(0, 1))
+        ...     with numpyro.plate("N", data.shape[0], subsample_size=100):
+        ...         batch = numpyro.subsample(data, event_dim=0)
+        ...         numpyro.sample("obs", dist.Normal(x, 1), obs=batch)
+        ...
+        >>> data = random.normal(random.PRNGKey(0), (10000,)) + 1
+        >>> gibbs_fn = subsample_gibbs_fn(model, data)
+        >>> kernel = HMCGibbs(NUTS(model), gibbs_fn, gibbs_sites=["N"])
+        >>> mcmc = MCMC(kernel, 1000, 1000)
+        >>> mcmc.run(random.PRNGKey(0), data)
+        >>> samples = mcmc.get_samples()["x"]
+        >>> assert abs(jnp.mean(samples).copy() - 1.) < 0.1
+
+    """
+    prototype_trace = trace(seed(model, rng_seed=0)).get_trace(*model_args, **model_kwargs)
+    plate_sizes = {
+        name: site["args"]
+        for name, site in prototype_trace.items()
+        if site["type"] == "plate" and site["args"][0] > site["args"][1]  # i.e. size > subsample_size
+    }
+    enum = any(site["type"] == "sample"
```

Comment:
this is used in a few places. perhaps move to a helper in infer/util?

Reply:
I think I will update `discrete_gibbs_fn` to not use this logic. Currently, `support_sizes` already holds all enumerable sites. We don't need to recompute this information in `discrete_gibbs_fn.gibbs_fn`